### PR TITLE
Allow canonical copy constructor/assignment argument (const &) for DualNumberSurrogate

### PR DESCRIPTION
**Pull request by lindsayad**
Without copy constructor/assignment operator that accepts a `const` argument, instantiations of stl containers that contain `DualNumberSurrogates` can error out. I've seen this in MOOSE work. I'm not quite sure why these errors aren't produced in the `nd_derivs_unit` test because it instantiates a `std::map` with `DualNumberSurrogate` values.

### CODE REVIEW

File: src/numerics/include/metaphysicl/dualnumber_surrogate.h [hunk 1]
```diff
@@ -57,17 +57,33 @@ inline DualNumberSurrogate<T, D>::DualNumberSurrogate(DualNumber<T2, D2> && dn,
 }
 
 template <typename T, typename D>
-inline DualNumberSurrogate<T, D>::DualNumberSurrogate(DualNumberSurrogate<T, D> & dns)
-  : _value(dns.value()), _derivatives(dns.derivatives())
+inline DualNumberSurrogate<T, D>::DualNumberSurrogate(const DualNumberSurrogate<T, D> & dns)
+  : _value(dns._value), _derivatives(dns._derivatives)
```
Comment (roystgnr) [inline src/numerics/include/metaphysicl/dualnumber_surrogate.h]: Why these changes?

File: src/numerics/include/metaphysicl/dualnumber_surrogate.h [hunk 2]
```diff
@@ -57,17 +57,33 @@ inline DualNumberSurrogate<T, D>::DualNumberSurrogate(DualNumber<T2, D2> && dn,
...
+  for (decltype(size) i = 0; i < size; ++i)
+  {
+    if (dns.derivatives()[i])
+      *_derivatives[i] = *dns.derivatives()[i];
+    else
+      *_derivatives[i] = 0;
+  }
+  return *this;
+}
+
```
Comment (roystgnr) [inline src/numerics/include/metaphysicl/dualnumber_surrogate.h]: If this works, why do we need a separate version accepting non-const?

File: src/numerics/include/metaphysicl/dualnumber_surrogate.h (hunk 1 above)
Comment (roystgnr) [reply src/numerics/include/metaphysicl/dualnumber_surrogate.h]: Replacing accessors with raw members, specifically.  Do we not have const versions of the accessors?

File: src/numerics/include/metaphysicl/dualnumber_surrogate.h [hunk 3]
```diff
@@ -57,17 +57,33 @@ inline DualNumberSurrogate<T, D>::DualNumberSurrogate(DualNumber<T2, D2> && dn,
...
-  : _value(dns.value()), _derivatives(dns.derivatives())
+inline DualNumberSurrogate<T, D>::DualNumberSurrogate(const DualNumberSurrogate<T, D> & dns)
+  : _value(dns._value), _derivatives(dns._derivatives)
 {
 }
 
 template <typename T, typename D>
 inline DualNumberSurrogate<T, D>::DualNumberSurrogate(DualNumberSurrogate<T, D> && dns)
-  : _value(dns.value()), _derivatives(dns.derivatives())
+  : _value(dns._value), _derivatives(dns._derivatives)
```
Comment (roystgnr) [inline src/numerics/include/metaphysicl/dualnumber_surrogate.h]: Same question here as above but without the same clarification; I don't have a guess at all...

**Comment by lindsayad** [comment]
How about this instead...
- more `const` instead of less. 
- Because at least one super common map type (`std::map`) isn't really meant to work with non-copy constructible objects (and `DNSSurrogate` as your comments drill down to  is not meant to be `const` copy-constructible) , use a `shared_ptr` to the `DNSSurrogates` instead

I know you didn't want to include `map`, but I'm hoping that `memory` is ok. I could use raw pointers, but then I would have to manually destruct everything

**Comment by roystgnr** [comment]
At first glance I think this version works for me.  Do you need it merged ASAP or could I take a harder look next week when I have more time?

**Comment by lindsayad** [comment]
I can wait.

**Comment by friedmud** [comment]
@roystgnr The importance of this is heating up a bit.  The PR to add automatic differentiation into MOOSE is almost ready to go: https://github.com/idaholab/moose/pull/12111 .  But we would like to have this change in before we go merging that.

File: src/numerics/include/metaphysicl/dualnumber_surrogate.h [hunk 4]
```diff
@@ -69,9 +69,8 @@ inline DualNumberSurrogate<T, D>::DualNumberSurrogate(DualNumberSurrogate<T, D>
 }
 
 template <typename T, typename D>
-template <typename T2, typename D2>
 inline DualNumberSurrogate<T, D> &
-DualNumberSurrogate<T, D>::operator=(DualNumberSurrogate<T2, D2> & dns)
```
Comment (roystgnr) [inline src/numerics/include/metaphysicl/dualnumber_surrogate.h]: Do we really need to lose the T2/D2 flexibility here?

**Comment by roystgnr** [comment]
Sorry about the delay.  I think that one question is the last of my remaining confusion, so whether the answer is "Yes, because" or "No, let's fix that" I'll be happy merging afterward.

File: src/numerics/include/metaphysicl/dualnumber_surrogate.h (hunk 4 above)
Comment (lindsayad) [reply src/numerics/include/metaphysicl/dualnumber_surrogate.h]: We're not losing it; see the next method definition. For some reason, before I had separate methods, one which took `const DualNumberSurrogate<T2,D2> &`, and another which took non-const `DualNumberSurrogate<T2,D2> &`. 

What this does is replace the extraneous `DualNumberSurrogate<T2,D2> &` with a user-defined copy assignment operator

Comment (roystgnr) [reply src/numerics/include/metaphysicl/dualnumber_surrogate.h]: Ah, I see now.  Thanks!

**Comment by roystgnr** [comment]
This isn't an important enough upgrade for a new release, right?  So we should just update the libMesh submodule to MetaPhysicL HEAD?

**Comment by lindsayad** [comment]
I agree that it's not worthy of an update all on it's own. Yea upgrade to a
boostrapped HEAD...?

**Comment by roystgnr** [comment]
Yeah; I'll make a new branch for the bootstrap commits and tag for the result but won't make it a release tag.

**Comment by roystgnr** [comment]
See if https://github.com/roystgnr/MetaPhysicL/releases/tag/pr19_bootstrapped works for you?

**Comment by lindsayad** [comment]
Yep, this works!
